flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchUpsertTableSinkBase.java

```diff
@@ -431,7 +438,8 @@ public ElasticsearchUpsertSinkFunction(
 				SerializationSchema<Row> serializationSchema,
 				XContentType contentType,
 				RequestFactory requestFactory,
-				int[] keyFieldIndices) {
+				int[] keyFieldIndices,
+				IndexFormatter indexFormatter) {
```

Comment:
I would suggest to accept a `IndexGenerator` interface, not a `IndexFormatter`.  `IndexFormatter` is just a factory to create `IndexGenerator`, even can be called `IndexGeneratorFactory`.  So that, in the future, DataStream users can also customize their special index generator. But can also use `IndexGeneratorFactory` in their DataStream jobs. 
```java
interface IndexGenerator extends Serializable {
  String generate(Row row);
}
```

Reply:
Thanks very much for you comments,  I'll update ASAP.